File: frontend/src/api/voucherTemplates.ts
import api from './client'
import type {
  DocumentType, VoucherTemplate, VoucherTemplateLine,
  GenerateDraftResponse, ListResponse,
} from '../types/invoice'

// ── Document Types ──────────────────────────────────────────

export async function getDocumentTypes(enabledOnly = false): Promise<DocumentType[]> {
  const res = await api.get<ListResponse<DocumentType>>('/document-types', {
    params: { enabled_only: enabledOnly },
  })
  return res.data.items
}

export async function createDocumentType(data: Partial<DocumentType>): Promise<DocumentType> {
  const res = await api.post<DocumentType>('/document-types', data)
  return res.data
}

export async function updateDocumentType(id: string, data: Partial<DocumentType>): Promise<DocumentType> {
  const res = await api.put<DocumentType>(`/document-types/${id}`, data)
  return res.data
}

export async function deleteDocumentType(id: string): Promise<void> {
  await api.delete(`/document-types/${id}`)
}

export async function restorePresetDocumentTypes(): Promise<DocumentType[]> {
  const res = await api.post<ListResponse<DocumentType>>('/document-types/restore-presets')
  return res.data.items
}

// ── Voucher Templates ───────────────────────────────────────

export async function getVoucherTemplates(documentTypeId?: string): Promise<VoucherTemplate[]> {
  const res = await api.get<ListResponse<VoucherTemplate>>('/voucher-templates', {
    params: documentTypeId ? { document_type_id: documentTypeId } : {},
  })
  return res.data.items
}

export async function getVoucherTemplate(id: string): Promise<VoucherTemplate> {
  const res = await api.get<VoucherTemplate>(`/voucher-templates/${id}`)
  return res.data
}

export async function createVoucherTemplate(data: {
  document_type_id?: string | null
  document_name: string
  settlement_method: string
  business_type: string
  summary_template: string
  priority?: number
  lines: {
    line_no: number
    debit_credit: string
    account_code: string
    account_name: string
    account_full_name?: string | null
    parent_account_code?: string | null
    amount_source: string
    require_sub_account?: boolean
    sub_account_match_mode?: string
    allow_manual_edit?: boolean
  }[]
}): Promise<VoucherTemplate> {
  const res = await api.post<VoucherTemplate>('/voucher-templates', data)
  return res.data
}

export async function updateVoucherTemplate(
  id: string,
  data: Partial<{
    document_type_id: string | null
    document_name: string
    settlement_method: string
    business_type: string
    summary_template: string
    is_enabled: boolean
    priority: number
    lines: any[]
  }>
): Promise<VoucherTemplate> {
  const res = await api.put<VoucherTemplate>(`/voucher-templates/${id}`, data)
  return res.data
}

export async function deleteVoucherTemplate(id: string): Promise<void> {
  await api.delete(`/voucher-templates/${id}`)
}

export async function copyVoucherTemplate(id: string): Promise<VoucherTemplate> {
  const res = await api.post<VoucherTemplate>(`/voucher-templates/${id}/copy`)
  return res.data
}

export async function toggleVoucherTemplate(id: string): Promise<VoucherTemplate> {
  const res = await api.put<VoucherTemplate>(`/voucher-templates/${id}/toggle`)
  return res.data
}

// ── Matching & Generation ───────────────────────────────────

export async function matchTemplates(params: {
  document_type_id?: string
  settlement_method?: string
  business_type?: string
  search_text?: string
}): Promise<{
  matched_templates: VoucherTemplate[]
  suggested_document_type_id: string | null
  suggested_business_type: string | null
  suggested_settlement_method: string | null
}> {
  const res = await api.post('/voucher-templates/match', params)
  return res.data
}

export async function generateDraftFromTemplate(data: {
  template_id: string
  client_id: string
  voucher_date?: string
  amounts: {
    total_amount?: number
    amount?: number
    tax_amount?: number
    income_amount?: number
    expense_amount?: number
    balance?: number
  }
  summary_vars?: Record<string, string>
  counterparty_name?: string
  source_invoice_id?: string
  source_transaction_id?: string
}): Promise<GenerateDraftResponse> {
  const res = await api.post<GenerateDraftResponse>('/voucher-templates/generate-draft', data)
  return res.data
}

File: backend/app/routers/voucher_templates.py
"""DocumentType + VoucherTemplate API routes."""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.voucher_template import (
    DocumentTypeCreate, DocumentTypeUpdate, DocumentTypeResponse, DocumentTypeListResponse,
    VoucherTemplateCreate, VoucherTemplateUpdate,
    VoucherTemplateResponse, VoucherTemplateListResponse,
    GenerateDraftRequest, GenerateDraftResponse,
    MatchTemplatesRequest, MatchTemplatesResponse,
    AmountData,
)
from app.services import voucher_template_engine as engine
from app.services.entry_service import create_entry

router = APIRouter(prefix="/api/v1", tags=["voucher-templates"])


# ═══════════════════════════════════════════════════════════════
# Document Types
# ═══════════════════════════════════════════════════════════════

@router.get("/document-types", response_model=DocumentTypeListResponse)
async def list_document_types(
    enabled_only: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    items = await engine.list_document_types(db, enabled_only=enabled_only)
    return DocumentTypeListResponse(
        items=[DocumentTypeResponse.model_validate(dt) for dt in items],
        total=len(items),
    )


@router.post("/document-types", response_model=DocumentTypeResponse, status_code=201)
async def create_document_type(data: DocumentTypeCreate, db: AsyncSession = Depends(get_db)):
    # Check duplicate code
    existing = await engine.get_document_type_by_code(db, data.code)
    if existing:
        raise HTTPException(status_code=400, detail=f"编码 {data.code} 已存在")
    dt = await engine.create_document_type(db, data)
    return DocumentTypeResponse.model_validate(dt)


@router.put("/document-types/{dt_id}", response_model=DocumentTypeResponse)
async def update_document_type(
    dt_id: str, data: DocumentTypeUpdate, db: AsyncSession = Depends(get_db)
):
    dt = await engine.get_document_type(db, dt_id)
    if not dt:
        raise HTTPException(status_code=404, detail="单据类别不存在")
    # Check code uniqueness if changing
    if data.code and data.code != dt.code:
        existing = await engine.get_document_type_by_code(db, data.code)
        if existing:
            raise HTTPException(status_code=400, detail=f"编码 {data.code} 已存在")
    dt = await engine.update_document_type(db, dt, data)
    return DocumentTypeResponse.model_validate(dt)


@router.delete("/document-types/{dt_id}", status_code=204)
async def delete_document_type(dt_id: str, db: AsyncSession = Depends(get_db)):
    dt = await engine.get_document_type(db, dt_id)
    if not dt:
        raise HTTPException(status_code=404, detail="单据类别不存在")
    if dt.is_system:
        raise HTTPException(status_code=400, detail="系统预置单据不能删除，只能停用")
    try:
        await engine.delete_document_type(db, dt)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return None


@router.post("/document-types/restore-presets", response_model=DocumentTypeListResponse)
async def restore_preset_document_types(db: AsyncSession = Depends(get_db)):
    """恢复预置单据类型：只补齐缺失的预置单据，不覆盖用户自定义"""
    restored = await engine.restore_preset_document_types(db)
    all_items = await engine.list_document_types(db)
    return DocumentTypeListResponse(
        items=[DocumentTypeResponse.model_validate(dt) for dt in all_items],
        total=len(all_items),
    )


# ═══════════════════════════════════════════════════════════════
# Voucher Templates
# ═══════════════════════════════════════════════════════════════

@router.get("/voucher-templates", response_model=VoucherTemplateListResponse)
async def list_voucher_templates(
    enabled_only: bool = Query(False),
    document_type_id: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    templates = await engine.list_templates(db, enabled_only=enabled_only)
    if document_type_id:
        templates = [t for t in templates if t.document_type_id == document_type_id]
    return VoucherTemplateListResponse(
        items=[VoucherTemplateResponse.model_validate(t) for t in templates],
        total=len(templates),
    )


@router.get("/voucher-templates/{tpl_id}", response_model=VoucherTemplateResponse)
async def get_voucher_template(tpl_id: str, db: AsyncSession = Depends(get_db)):
    tpl = await engine.get_template(db, tpl_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")
    return VoucherTemplateResponse.model_validate(tpl)


@router.post("/voucher-templates", response_model=VoucherTemplateResponse, status_code=201)
async def create_voucher_template(data: VoucherTemplateCreate, db: AsyncSession = Depends(get_db)):
    lines_data = [ld.model_dump() for ld in data.lines]
    tpl = await engine.create_template(
        db,
        document_name=data.document_name,
        settlement_method=data.settlement_method,
        business_type=data.business_type,
        summary_template=data.summary_template,
        lines_data=lines_data,
        document_type_id=data.document_type_id,
        company_id=data.company_id,
        priority=data.priority,
    )
    # Re-fetch for relationships
    tpl = await engine.get_template(db, tpl.id)
    return VoucherTemplateResponse.model_validate(tpl)


@router.put("/voucher-templates/{tpl_id}", response_model=VoucherTemplateResponse)
async def update_voucher_template(
    tpl_id: str, data: VoucherTemplateUpdate, db: AsyncSession = Depends(get_db)
):
    tpl = await engine.get_template(db, tpl_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")
    update_data = data.model_dump(exclude_unset=True)
    if "lines" in update_data and update_data["lines"]:
        update_data["lines"] = [ld if isinstance(ld, dict) else ld.model_dump()
                                for ld in update_data["lines"]]
    tpl = await engine.update_template(db, tpl, update_data)
    tpl = await engine.get_template(db, tpl.id)
    return VoucherTemplateResponse.model_validate(tpl)


@router.delete("/voucher-templates/{tpl_id}", status_code=204)
async def delete_voucher_template(tpl_id: str, db: AsyncSession = Depends(get_db)):
    tpl = await engine.get_template(db, tpl_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")
    await engine.delete_template(db, tpl)
    return None


@router.post("/voucher-templates/{tpl_id}/copy", response_model=VoucherTemplateResponse, status_code=201)
async def copy_voucher_template(tpl_id: str, db: AsyncSession = Depends(get_db)):
    tpl = await engine.get_template(db, tpl_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")
    new_tpl = await engine.copy_template(db, tpl)
    new_tpl = await engine.get_template(db, new_tpl.id)
    return VoucherTemplateResponse.model_validate(new_tpl)


@router.put("/voucher-templates/{tpl_id}/toggle", response_model=VoucherTemplateResponse)
async def toggle_voucher_template(tpl_id: str, db: AsyncSession = Depends(get_db)):
    tpl = await engine.get_template(db, tpl_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")
    tpl = await engine.toggle_template(db, tpl)
    return VoucherTemplateResponse.model_validate(tpl)


# ═══════════════════════════════════════════════════════════════
# Matching & Generation
# ═══════════════════════════════════════════════════════════════

@router.post("/voucher-templates/match", response_model=MatchTemplatesResponse)
async def match_templates_endpoint(
    data: MatchTemplatesRequest, db: AsyncSession = Depends(get_db)
):
    """根据单据类型/结算方式/业务类型匹配分录模板"""
    # Try to identify business type from search text
    biz_type = data.business_type
    if not biz_type and data.search_text:
        biz_type, _ = engine.identify_business_type(data.search_text)

    matched = await engine.match_templates(
        db,
        document_type_id=data.document_type_id,
        settlement_method=data.settlement_method,
        business_type=biz_type,
    )

    return MatchTemplatesResponse(
        matched_templates=[VoucherTemplateResponse.model_validate(t) for t in matched],
        suggested_document_type_id=data.document_type_id,
        suggested_business_type=biz_type,
        suggested_settlement_method=data.settlement_method,
    )


@router.post("/voucher-templates/generate-draft", response_model=GenerateDraftResponse)
async def generate_draft(
    data: GenerateDraftRequest, db: AsyncSession = Depends(get_db)
):
    """根据模板+金额数据生成凭证草稿预览"""
    tpl = await engine.get_template(db, data.template_id)
    if not tpl:
        raise HTTPException(status_code=404, detail="分录模板不存在")

    voucher_date = None
    if data.voucher_date:
        try:
            voucher_date = date.fromisoformat(data.voucher_date)
        except ValueError:
            raise HTTPException(status_code=400, detail="日期格式无效，请使用 YYYY-MM-DD")

    draft = await engine.generate_draft_from_template(
        db, tpl, data.amounts, data.client_id,
        voucher_date=voucher_date,
        summary_vars=data.summary_vars,
        counterparty_name=data.counterparty_name,
        source_invoice_id=data.source_invoice_id,
    )

    # If no errors, create the actual entry
    if not draft.errors:
        entry_data = await engine.create_draft_entry(
            db, tpl, data.amounts, data.client_id,
            voucher_date=voucher_date,
            summary_vars=data.summary_vars,
            counterparty_name=data.counterparty_name,
            source_invoice_id=data.source_invoice_id,
        )
        entry = await create_entry(db, entry_data)
        draft.entry_id = entry.id

    return draft
